src/compiler/program.ts

```diff
@@ -1100,6 +1121,24 @@ namespace ts {
             };
         }
 
+        function getPrependNodes(): PrependNode[] {
+            if (!options.references) {
+                return emptyArray;
+            }
+
+            const nodes: PrependNode[] = [];
+            walkProjectReferenceGraph(host, options, (_file, proj, opts) => {
+                if (opts.prepend) {
+                    const dtsFilename = changeExtension(proj.outFile, ".d.ts");

```

Comment:
This needs to account for `declarationDir`, and/or `outDir`, right?

Reply:
`outFile` and `declarationDir` cannot be combined; ref #12405

`outFile` with `outDir` - doesn't matter; the output .d.ts file is dropped next to the output .js file

Need an error for a prepended reference that doesn't use `outFile`, though

Reply:
✔️